ISISScientificComputing/autoreduce
# Removes variable duplication when submitting runs

### Summary of work
Does a big refactor of `queue_processors/queue_processor/queueproc_utils/instrument_variable_utils.py`. The Web App copy of the same code will be handled in another PR

Effects of this PR:
Variables should no longer be duplicated.

### How to test your work
To test you'll have to deploy the migrated MySQL changes while running the DB locally. Then update your credentials accordingly.

Throughout this I recommend checking the number of rows in `InstrumentVariables` and `Variables`.

- Submit GEM 89880
	- Check that 5 variables were created as per it's reduce_vars.py
- Submit MARI 27858
	- Check that 10 variables have been created
- Submit GEM 89881
	- Check that no new variables have been created, but all have been re-used
- Change GEM's reduce_vars value, and resubmit
	- One new variable should have been created, containing the new value
	- Revert GEM's variable value
- Submit GEM 89881 again and MARI 27858
- Variable count shouldn't have changed, all should have been re-used

### Additional comments

Note: this requires a database migration due to the change in `RunVariable` (making `InstrumentVariable` a FK relation, to allow re-use).

Fixes #285
